### PR TITLE
Adding detection of IP address with ifcfg==0.11b6

### PR DESCRIPTION
**Pull request by benjaoming**
## Summary

Please see my upstream changes (>1000 lines of code):

https://github.com/ftao/python-ifcfg/pull/7
https://github.com/ftao/python-ifcfg/pull/8

## TODO

- [x] Have tests been written for the new code? (sort of, ifcfg is really well-tested)

## Reviewer guidance

Please take over the PR in case issues come up - or feel free to mess with the code due to my absence the next 2 weeks.

## Issues addressed

#1595

## Screenshots (if appropriate)

image

### CODE REVIEW

File: kolibri/utils/cli.py [hunk 1]
```diff
@@ -235,9 +235,24 @@ def start(port=None, daemon=True):
...
     else:
         logger.info("Running 'kolibri start' as daemon (system service)")
 
+    __, urls = server.get_urls(listen_port=port)
+    if not urls:
+        logger.error(
+            "Could not detect an IP address that Kolibri binds to, but try "
+            "opening up the following addresses:\n")
+        urls = [
+            "http://{}:{}".format(ip, port) for ip in ("0.0.0.0", "127.0.0.1")
```
Comment (rtibbles) [inline kolibri/utils/cli.py]: Per discussion in the dev meeting, we might want to remove any references to `0.0.0.0`. Also, may be worth emphasizing a distinction between local and network IPs for less IP savvy users?

**Comment by rtibbles** [comment]
Will wait on testing feedback from @radinamatic, and then make the small tweak I suggest before merging.

File: kolibri/utils/cli.py (hunk 1 above)
Comment (benjaoming) [reply kolibri/utils/cli.py]: You're probably right - there's some more definitions of it here: https://en.wikipedia.org/wiki/0.0.0.0

Notice the text above should make it clear to the user that the IPs may not work.

What I was mostly judging from was simply that 0.0.0.0 resolves to localhost, you can also use http://0:8080 as a short.. but dunno if it's just Firefox that resolves it and redirects.

Feel free to call the shots! But the intention is to make this decision as irrelevant as possible :)

**Comment by radinamatic** [comment]
On Windows 7 & 8 url detection seem to be working OK, :+1:. There are some additional urls detected - not sure if that can be polished, or those are due to network adaptes on VMs:

win7 start running - oracle vm virtualbox_426

However, Windows 10 has issues starting the server, and I see a bunch of `ifcfg` errors in the command prompt, so that might be the most likely culprit.

virtualbox_msedge - win10_preview_02_08_2017_19_50_39

Ah, and `kolibri url` command is not implemented, yet - for some reason I thought it was supposed to be in this PR. 

Anything else more specific that I need to test @rtibbles?

**Comment by rtibbles** [comment]
Looks good, @radinamatic - thanks, very thorough.

**Comment by rtibbles** [comment]
Seems like this might need some upstream work for Windows 10 compatibility?

**Comment by benjaoming** [comment]
@radinamatic can you send the raw output of `ipconfig` as run from the command line in this Win 10? Seems it contains characters that don't parse as UTF-8, but moreover there's a need for more investigation on how to handle the stdout of some command line output on *any* given system.

**Comment by radinamatic** [comment]
You should be enjoying Provance and herding sheep, not yearning raw command prompt outputs at this hour... :stuck_out_tongue: 

```
Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. All rights reserved.

C:\Users\IEUser>ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : MSEDGEWIN10
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No

Ethernet adapter Ethernet:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Intel(R) PRO/1000 MT Desktop Adapter
   Physical Address. . . . . . . . . : 08-00-27-CC-BE-AF
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::e59b:7afd:78a6:d073%4(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.0.2.15(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : jueves, 3 de agosto de 2017 1:00:44
   Lease Expires . . . . . . . . . . : viernes, 4 de agosto de 2017 1:00:59
   Default Gateway . . . . . . . . . : 10.0.2.2
   DHCP Server . . . . . . . . . . . : 10.0.2.2
   DHCPv6 IAID . . . . . . . . . . . : 34078759
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1F-22-AC-9F-08-00-27-CC-BE-AF
   DNS Servers . . . . . . . . . . . : 89.150.129.22
                                       89.150.129.10
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Ethernet 2:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Intel(R) PRO/1000 MT Desktop Adapter #2
   Physical Address. . . . . . . . . : 08-00-27-0D-9A-0B
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::99e0:2790:ba60:20e1%5(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.56.101(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : jueves, 3 de agosto de 2017 1:00:44
   Lease Expires . . . . . . . . . . : jueves, 3 de agosto de 2017 1:20:59
   Default Gateway . . . . . . . . . :
   DHCP Server . . . . . . . . . . . : 192.168.56.100
   DHCPv6 IAID . . . . . . . . . . . : 151519271
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1F-22-AC-9F-08-00-27-CC-BE-AF
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Tunnel adapter isatap.{ADDEF65B-69BD-40F2-A0E6-4B67361ECE85}:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft ISATAP Adapter
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Tunnel adapter isatap.{BFDBE788-14F2-499A-9D83-CDCD34CDE463}:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft ISATAP Adapter #2
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

C:\Users\IEUser>ipconfig /all | clip

C:\Users\IEUser>
```

**Comment by benjaoming** [comment]
@radinamatic the latest Buildkite assets are available for testing now - could you check if the encoding error on Windows has disappeared?

**Comment by radinamatic** [comment]
@benjaoming Sure, downloading now! :+1:

**Comment by radinamatic** [comment]
All is well on Win10 front @benjaoming :100: 

w10edge start running - oracle vm virtualbox_471

**Comment by benjaoming** [comment]
Just wanted to share that I learned this insight about coverage reports. When it says:

> Merging [#1929](https://codecov.io/gh/learningequality/kolibri/pull/1929?src=pr&el=desc) into [release-v0.5.x](https://codecov.io/gh/learningequality/kolibri/commit/0570b5772884148fa5e8de89a8da9fddec82b34c?src=pr&el=desc) will **decrease** coverage by `0.02%`.

Notice that it references 0570b5772884148fa5e8de89a8da9fddec82b34c - meaning that it does the comparison to an older absolute metric.

My patch actually improves coverage (as you can read from the relative delta) - this often baffled me.

I have tried rebasing before, but as I recall, it doesn't change that CodeCove keeps comparing to the initial commit hash that the target branch had when opening the PR.

Coverage went up from 73.81% to 73.98% :)

**Pull request merged**